### PR TITLE
fix: use epoch instead of u64::MAX - epoch

### PR DESCRIPTION
**Pull request by xinchengxx**
Signed-off-by: xinchengxx <2240131203@qq.com>

I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

As the [issue5936](https://github.com/risingwavelabs/risingwave/issues/5963) mentioned, currently our fullkey's length is not fixed, so we can't use the u64::MAX - epoch to sort the fullkey by dictionary order. we can just append the epoch to the last 8 bytes.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
[issue5936](https://github.com/risingwavelabs/risingwave/issues/5963)

### CODE REVIEW

**Review by skyzh — COMMENTED** [review]
You might also need to look into the comparators of FullKey? So that they're comparing in a correct way.

**Comment by Gun9niR** [comment]
See `VersionedComparator`.

**Comment by xinchengxx** [comment]
But i still need to review some unit tests, because of the `next_key` and `prev_key` function will never work during other unit tests for generator seek key if we changed the `u64::max - epoch` to `epoch`

**Comment by skyzh** [comment]
> generating seek key if we changed the u64::max - epoch to epoch

If generating seek key is done by something like FullKey(key, MAX_EPOCH), this should never happen. Probably you'll need to do some refactors to make the semantics more clear.

File: src/storage/hummock_sdk/src/key.rs [hunk 1]
```diff
@@ -153,6 +151,53 @@ fn next_key_no_alloc(key: &[u8]) -> Option<(&[u8], u8)> {
     Some((&key[..pos], key[pos] + 1))
 }
 
+/// compute the next full key of the given full key
+///
+/// if the `user_key` has no successor key, the result will be a empty vec
+
+pub fn next_full_key(full_key: &[u8]) -> Vec<u8> {
+    let (user_key, epoch) = split_key_epoch(full_key);
+    let next_epoch = prev_key(epoch);
```
Comment (Li0k) [inline src/storage/hummock_sdk/src/key.rs:160]: use `prev_key` for `next_epoch` is a little comfused.

add some notes or use another function for getting  `next_epoch`

File: src/storage/hummock_sdk/src/key.rs [hunk 2]
```diff
@@ -153,6 +151,53 @@ fn next_key_no_alloc(key: &[u8]) -> Option<(&[u8], u8)> {
...
+    }
+}
+
+/// compute the prev full key of the given full key
+///
+/// if the `user_key` has no predecessor key, the result will be a empty vec
+
+pub fn prev_full_key(full_key: &[u8]) -> Vec<u8> {
+    let (user_key, epoch) = split_key_epoch(full_key);
+    let mut prev_epoch = next_key(epoch);
```
Comment (Li0k) [inline src/storage/hummock_sdk/src/key.rs:183]: Consistent with the comments above

Comment (zwang28) [inline src/storage/hummock_sdk/src/key.rs:183]: next_key(epoch) may return invalid prev_epoch (less bytes), e.g.
`assert_eq!(next_key(b"\xff\x00\xff"), b"\xff\x01");`

**Review by Li0k — APPROVED** [review]
LGTM , PTAL @zwang28 @Little-Wallace

**Review by Little-Wallace — APPROVED** [review]
LGTM

**Pull request merged**